cylc/flow/task_proxy.py

```diff
@@ -355,25 +355,37 @@ def get_point_as_seconds(self):
                 self.point_as_seconds += utc_offset_in_seconds
         return self.point_as_seconds
 
-    def get_clock_trigger_time(self, offset_str):
-        """Compute, cache, and return trigger time relative to cycle point.
+    def get_clock_trigger_time(
+        self,
+        point: 'PointBase', offset_str: Optional[str] = None
+    ) -> int:
+        """Compute, cache and return trigger time relative to cycle point.
 
         Args:
-            offset_str: ISO8601Interval string, e.g. "PT2M".
-                        Can be None for zero offset.
+            point:
+                String representing itask string.
+            offset_str:
+                ISO8601Interval string, e.g. "PT2M".
+                Can be None for zero offset.
         Returns:
             Absolute trigger time in seconds since Unix epoch.
 
         """
-        if self.clock_trigger_time is None:
-            if offset_str is None:
-                trigger_time = self.point
+        # None cannot be used as a dict key:
+        offset_str = offset_str if offset_str else 'P0Y'
```

Comment:
Yes, although I've never tried using `None` as a dictionary key before, and was somewhat surprised by it as a thing.

Reply:
It's a bit more complex in practice - because if `offset_str` _is_ in `self.clock_trigger_times` it makes the following logic invalid and you have to write a separate loop caching the conversion of `P0Y` into seconds since epoch. I ended up with this diff: 

```diff
-        # None cannot be used as a dict key:
-        offset_str = offset_str if offset_str else 'P0Y'
-        if offset_str not in self.clock_trigger_times:
-            if offset_str == 'P0Y':
-                trigger_time = point
-            else:
-                trigger_time = point + ISO8601Interval(offset_str)
-
+        trigger_time = None
+        if (
+            offset_str is None
+            and self.clock_trigger_times[offset_str] == 'P0Y'
+        ):
+            trigger_time = point
+        elif offset_str not in self.clock_trigger_times:
+            trigger_time = point + ISO8601Interval(offset_str)
+        if trigger_time:

```

Reply:
`None` is hashable so can be used as a dict key

Reply:
> `None` is hashable so can be used as a dict key

Yes. TIL - still doesn't deal with the other problem.

Reply:
Too late - Oliver's merged it!